src/components/user.js

```diff
@@ -50,34 +97,66 @@ export default class User extends React.Component {
 
             UserAction.updateUser(newUser);
         }
-
         this.setState({ wantEdit: !wantEdit });
         e.preventDefault();
     }
 
-    handleChange (e, type) {
+
+    handleChange (e, obj) {
         const { currentUser, wantEdit } = this.state;
-        const {isNewUser} = this.props;
+        const { isNewUser }             = this.props;
 
         if (wantEdit || isNewUser) {
-            currentUser[type][e.target.name] = e.target.value;
-            this.setState({ currentUser });
+            const name  = e.target.name;
+            const value = e.target.value;
+
+            currentUser[obj][name] = value;
+            this.setState({ currentUser }, () => {
+                this._validateField(name, value);
+            });
         }
         e.preventDefault();
     }
 
-    handleClickSubmit () {
-        const { currentUser } = this.state;
-        const newUser         = new UserInfo(currentUser.info, currentUser.address, currentUser.company);
+    handleClickSubmit (e) {
+        const { currentUser, addDuplicate, formErrors } = this.state;
+        const newUser                                   = new UserInfo(currentUser.info, currentUser.address, currentUser.company);
 
-        UserAction.addNewUser(newUser);
+        if (UserStore.checkUserNameExist(newUser) && !addDuplicate) {
+            this.setState({ addDuplicate: !addDuplicate });
+            const userId = UserStore.getUserIdExist(newUser);
+
+            formErrors.dublicate = 'User with this name exists. Click Submit if you want to add anyway.';
+            formErrors.href      = `#${userId}`;
+            this.setState({
+                formErrors: formErrors
+            });
+        }
+        else {
+            const validName  = this._validateField('name', currentUser.info.name);
```

Comment:
we will have three `setState` methods calling here.
Let's our `_validateField` just return the result of the checking and we will call the `setState` in the caller method (merged with the its own changes).
We can even impove logic of this checkiing^
- we can have two methods `_validateField` and `_validateFields`. The second one will get an array, call the first one and return some usefull object with information about checking
- or we can giva a capability pass in the `_validateField` separate value or an array